File: src/edu/stanford/cfuller/imageanalysistools/filter/ConvolutionFilter.java
/* ***** BEGIN LICENSE BLOCK *****
 * 
 * Copyright (c) 2011 Colin J. Fuller
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * 
 * ***** END LICENSE BLOCK ***** */

package edu.stanford.cfuller.imageanalysistools.filter;

import org.apache.commons.math.complex.Complex;
import org.apache.commons.math.transform.FastFourierTransformer;

import edu.stanford.cfuller.imageanalysistools.image.Histogram;
import edu.stanford.cfuller.imageanalysistools.image.Image;
import edu.stanford.cfuller.imageanalysistools.image.ImageCoordinate;

/**
 * Convolves a 2D x-y Image with a 2D kernel.
 * <p>
 * Before calling apply, the kernel must be specified using the {@link #setKernel(Kernel)} method.
 * <p>
 * This filter does not use a reference Image.
 * <p>
 * After calling the apply method, the image argument to apply will be overwritten by the convolved Image.
 * 
 * @author Colin J. Fuller
 */
public class ConvolutionFilter extends Filter {

	//TODO: deal with boundary conditions of types other than zero.
	
	//TODO: deal with something other than single image plane transforms.
	
	Kernel k;
	
	Complex[][] transformStorage;
	
	@Override
	public void apply(Image im) {
		
		Complex[][] transformed = null;
		
		if (this.transformStorage != null) {
			transformed = new Complex[this.transformStorage.length][this.transformStorage[0].length];
			for (int i =0; i < transformed.length; i++) {
				for (int j = 0; j < transformed.length; j++) {
					transformed[i][j] = this.transformStorage[i][j];
				}
			}
			
		} else {
			transformed = ConvolutionFilter.transform(im, 0);

		}
				
		Complex[][] kernelTransform = this.k.getTransformed2DKernel(transformed[0].length, transformed.length);
				
		for (int i = 0; i < transformed.length; i++) {
			for (int j = 0; j < transformed.length; j++) {
				transformed[i][j] = transformed[i][j].multiply(kernelTransform[i][j]);
			}
		}
		
		inverseTransform(im,0,transformed);
		
	}
	
	public void setTransform(Complex[][] transformed) {
		this.transformStorage = transformed;
	}
	
	public void setKernel(Kernel k) {
		this.k = k;
	}
	
	
	public static Complex[][] transform(Image im, int z) {

        FastFourierTransformer fft = new org.apache.commons.math.transform.FastFourierTransformer();

        int ydimPowOfTwo = im.getDimensionSizes().get(ImageCoordinate.Y);
        int xdimPowOfTwo = im.getDimensionSizes().get(ImageCoordinate.X);

        if (!FastFourierTransformer.isPowerOf2(ydimPowOfTwo) || !FastFourierTransformer.isPowerOf2(xdimPowOfTwo)) {

            xdimPowOfTwo = (int) Math.pow(2, Math.ceil(Math.log(im.getDimensionSizes().get(ImageCoordinate.X)) / Math.log(2)));
            ydimPowOfTwo = (int) Math.pow(2, Math.ceil(Math.log(im.getDimensionSizes().get(ImageCoordinate.Y))/Math.log(2)));
        }

        //for (int p =0; p < im.getPlaneCount(); p++) {

        int p = z;
        
            im.selectPlane(p);

            double[][] rowImage = new double[ydimPowOfTwo][xdimPowOfTwo];
            for (int i =0; i < ydimPowOfTwo; i++) {
                java.util.Arrays.fill(rowImage[i], 0); // ensures zero-padding
            }
            Complex[][] colMajorImage = new Complex[xdimPowOfTwo][ydimPowOfTwo];

            for (ImageCoordinate ic : im) {
                rowImage[ic.get(ImageCoordinate.Y)][ic.get(ImageCoordinate.X)] = im.getValue(ic);
            }

            for (int r = 0; r < rowImage.length; r++) {
                double[] row = rowImage[r];
                Complex[] transformedRow = fft.transform(row);

                for (int c = 0; c < colMajorImage.length; c++) {
                    colMajorImage[c][r] = transformedRow[c];
                }
            }

            for (int c = 0; c < colMajorImage.length; c++) {
                colMajorImage[c] = fft.transform(colMajorImage[c]);
            }
            return colMajorImage;
        //}
	}
	
	public void inverseTransform(Image orig, int z, Complex[][] transformed) {
		
		Complex[][] colMajorImage = transformed;
		
		double[][] rowImage = new double[colMajorImage[0].length][colMajorImage.length];
		
        FastFourierTransformer fft = new org.apache.commons.math.transform.FastFourierTransformer();


        for (int c = 0; c < colMajorImage.length; c++) {
            colMajorImage[c] = fft.inversetransform(colMajorImage[c]);
        }

        Complex[] tempRow = new Complex[rowImage.length];

        //also calculate min/max values
        double newMin = Double.MAX_VALUE;
        double newMax = 0;

        for (int r = 0; r < rowImage.length; r++) {

            for (int c = 0; c < colMajorImage.length; c++) {
                tempRow[c] = colMajorImage[c][r];
            }

            Complex[] transformedRow = fft.inversetransform(tempRow);

            for (int c = 0; c < colMajorImage.length; c++) {
                rowImage[r][c] = transformedRow[c].abs();
                if (rowImage[r][c] < newMin) newMin = rowImage[r][c];
                if (rowImage[r][c] > newMax) newMax = rowImage[r][c];
            }
        }

        //rescale values to same min/max as before

        Histogram h = new Histogram(orig);

        double oldMin = h.getMinValue();
        double oldMax = h.getMaxValue();

        double scaleFactor = (oldMax - oldMin)/(newMax - newMin);

        ImageCoordinate minCoord = ImageCoordinate.createCoordXYZCT(0, 0, z, 0, 0);
        ImageCoordinate maxCoord = ImageCoordinate.cloneCoord(orig.getDimensionSizes());
        maxCoord.set(ImageCoordinate.Z,z+1);
        
        orig.setBoxOfInterest(minCoord, maxCoord);
        
        scaleFactor = 1;
        oldMin  = 0;

        for (ImageCoordinate ic : orig) {
            orig.setValue(ic, (float) ((rowImage[ic.get(ImageCoordinate.Y)][ic.get(ImageCoordinate.X)] - newMin)*scaleFactor + oldMin));
        }
        
        orig.clearBoxOfInterest();
        
        minCoord.recycle();
        maxCoord.recycle();


    
	}

}

File: src/main/java/edu/stanford/cfuller/imageanalysistools/filter/ConvolutionFilter.kt
package edu.stanford.cfuller.imageanalysistools.filter

import org.apache.commons.math3.complex.Complex
import org.apache.commons.math3.transform.FastFourierTransformer

import edu.stanford.cfuller.imageanalysistools.image.Histogram
import edu.stanford.cfuller.imageanalysistools.image.WritableImage
import edu.stanford.cfuller.imageanalysistools.image.Image
import edu.stanford.cfuller.imageanalysistools.image.ImageCoordinate

/**
 * Convolves a 2D x-y Image with a 2D kernel.
 *
 *
 * Before calling apply, the kernel must be specified using the [.setKernel] method.
 *
 *
 * This filter does not use a reference Image.
 *
 *
 * After calling the apply method, the image argument to apply will be overwritten by the convolved Image.

 * @author Colin J. Fuller
 */
class ConvolutionFilter : Filter() {
    // TODO: deal with boundary conditions of types other than zero.
    // TODO: deal with something other than single image plane transforms.

    internal var k: Kernel = Kernel.noOp()
    internal var transformStorage: Array<Array<Complex>>? = null

    override fun apply(im: WritableImage) {
        val tfStorage = this.transformStorage
        val transformed: Array<Array<Complex>> = if (tfStorage != null) {
            val tf = Array<Array<Complex>>(tfStorage.size) { i ->
                Array<Complex>(tfStorage[0].size) { j ->
                    tfStorage[i][j]
                }
            }
            tf
        } else {
            ConvolutionFilter.transform(im, 0)
        }

        val kernelTransform = this.k.getTransformed2DKernel(transformed[0].size, transformed.size)

        for (i in transformed.indices) {
            for (j in transformed.indices) {
                transformed[i][j] = transformed[i][j].multiply(kernelTransform[i][j])
            }
        }

        inverseTransform(im, 0, transformed)
    }

    fun setTransform(transformed: Array<Array<Complex>>) {
        this.transformStorage = transformed
    }

    fun setKernel(k: Kernel) {
        this.k = k
    }

    fun inverseTransform(orig: WritableImage, z: Int, transformed: Array<Array<Complex>>) {
        val colMajorImage = transformed
        val rowImage = Array(colMajorImage[0].size) { DoubleArray(colMajorImage.size) }
        val fft = org.apache.commons.math3.transform.FastFourierTransformer(org.apache.commons.math3.transform.DftNormalization.STANDARD)

        for (c in colMajorImage.indices) {
            colMajorImage[c] = fft.transform(colMajorImage[c], org.apache.commons.math3.transform.TransformType.INVERSE)
        }

        val tempRow = arrayOfNulls<Complex>(rowImage.size)

        //also calculate min/max values
        var newMin = java.lang.Double.MAX_VALUE
        var newMax = 0.0

        for (r in rowImage.indices) {

            for (c in colMajorImage.indices) {
                tempRow[c] = colMajorImage[c][r]
            }

            val transformedRow = fft.transform(tempRow, org.apache.commons.math3.transform.TransformType.INVERSE)

            for (c in colMajorImage.indices) {
                rowImage[r][c] = transformedRow[c].abs()
                if (rowImage[r][c] < newMin) newMin = rowImage[r][c]
                if (rowImage[r][c] > newMax) newMax = rowImage[r][c]
            }
        }

        //rescale values to same min/max as before

        val h = Histogram(orig)

        var oldMin = h.minValue.toDouble()
        val oldMax = h.maxValue.toDouble()

        var scaleFactor = (oldMax - oldMin) / (newMax - newMin)

        val minCoord = ImageCoordinate.createCoordXYZCT(0, 0, z, 0, 0)
        val maxCoord = ImageCoordinate.cloneCoord(orig.dimensionSizes)
        maxCoord[ImageCoordinate.Z] = z + 1

        orig.setBoxOfInterest(minCoord, maxCoord)

        scaleFactor = 1.0
        oldMin = 0.0

        for (ic in orig) {
            orig.setValue(ic, ((rowImage[ic[ImageCoordinate.Y]][ic[ImageCoordinate.X]] - newMin) * scaleFactor + oldMin).toFloat())
        }

        orig.clearBoxOfInterest()
        minCoord.recycle()
        maxCoord.recycle()
    }

    companion object {
        fun transform(im: Image, z: Int): Array<Array<Complex>> {
            val fft = org.apache.commons.math3.transform.FastFourierTransformer(org.apache.commons.math3.transform.DftNormalization.STANDARD)
            var ydimPowOfTwo = im.dimensionSizes[ImageCoordinate.Y]
            var xdimPowOfTwo = im.dimensionSizes[ImageCoordinate.X]

            if (!org.apache.commons.math3.util.ArithmeticUtils.isPowerOfTwo(ydimPowOfTwo.toLong()) || !org.apache.commons.math3.util.ArithmeticUtils.isPowerOfTwo(xdimPowOfTwo.toLong())) {
                xdimPowOfTwo = Math.pow(2.0, Math.ceil(Math.log(im.dimensionSizes[ImageCoordinate.X].toDouble()) / Math.log(2.0))).toInt()
                ydimPowOfTwo = Math.pow(2.0, Math.ceil(Math.log(im.dimensionSizes[ImageCoordinate.Y].toDouble()) / Math.log(2.0))).toInt()
            }

            val p = z
            im.selectPlane(p)

            val rowImage = Array(ydimPowOfTwo) { DoubleArray(xdimPowOfTwo) }
            for (i in 0..ydimPowOfTwo - 1) {
                java.util.Arrays.fill(rowImage[i], 0.0) // ensures zero-padding
            }
            for (ic in im) {
                rowImage[ic[ImageCoordinate.Y]][ic[ImageCoordinate.X]] = im.getValue(ic).toDouble()
            }
            val colMajorImage = Array<Array<Complex>>(xdimPowOfTwo) { Array<Complex>(ydimPowOfTwo, { Complex(0.0, 0.0) }) }

            for (r in rowImage.indices) {
                val row = rowImage[r]
                val transformedRow = fft.transform(row, org.apache.commons.math3.transform.TransformType.FORWARD)

                for (c in colMajorImage.indices) {
                    colMajorImage[c][r] = transformedRow[c]
                }
            }

            for (c in colMajorImage.indices) {
                colMajorImage[c] = fft.transform(colMajorImage[c], org.apache.commons.math3.transform.TransformType.FORWARD)
            }
            return colMajorImage
        }
    }
}
